### PR TITLE
Add unit testing capabilities for templates

### PR DESCRIPTION
**Pull request by DeepDiver1975**
With these changes unit testing templates can be done as easy as

``` php
    public function test403() {
        $template = \OC::$SERVERROOT . '/core/templates/403.php';
        $expectedHtml = "<ul><li class='error'>\n\t\tAccess forbidden<br><p class='hint'></p></li></ul>";
        $this->assertTemplate($expectedHtml, $template);
    }
```

### CODE REVIEW

File: lib/private/template/base.php [hunk 1]
```diff
@@ -131,29 +135,33 @@ public function printPage() {
...
-		return $this->load($this->template);
+	public function fetchPage($additionalParams = null) {
+		return $this->load($this->template, $additionalParams);
 	}
 
 	/**
 	 * doing the actual work
+	 *
 	 * @param string $file
+	 * @param null $additionalParams
```
Comment (LukasReschke) [inline lib/private/template/base.php]: Only null? :wink:

File: tests/lib/testcase.php [hunk 2]
```diff
@@ -408,4 +415,67 @@ private function IsDatabaseAccessAllowed() {
...
+		$expected->loadHTML($expectedHtml);
+
+		$actual = new DOMDocument();
+		$actual->preserveWhiteSpace = false;
+		$actual->formatOutput = true;
+		$actual->loadHTML($actualHtml);
+		$this->removeWhitespaces($actual);
+
+//		$expectedHtml1 = preg_replace('~\s+~i', '', $expected->saveHTML());
+//		$actualHtml1 = preg_replace('~\s+~i', '', $actual->saveHTML());
```
Comment (MorrisJobke) [inline tests/lib/testcase.php]: could be removed ;)

**Comment by MorrisJobke** [comment]
Code looks good and works 👍

File: tests/lib/testcase.php [hunk 3]
```diff
@@ -408,4 +415,65 @@ private function IsDatabaseAccessAllowed() {
 
 		return false;
 	}
+
+	/**
+	 * @param string $expectedHtml
+	 * @param string $template
+	 * @param array $vars
+	 */
+	protected function assertTemplate($expectedHtml, $template, $vars = []) {
```
Comment (nickvergessen) [inline tests/lib/testcase.php]: Can't those methods live in  `tests/core/templates/templates.php`

Comment (DeepDiver1975) [inline tests/lib/testcase.php]: I want to use this function in very app and all templates here in core - moving them into the base class made most sense to me - objections?

Comment (nickvergessen) [inline tests/lib/testcase.php]: Well my "objection" was that it is now available in all apps and tests. but I guess that is intended then, so fine

Comment (DeepDiver1975) [inline tests/lib/testcase.php]: > so fine

is that a :+1: :question:

**Comment by LukasReschke** [comment]
:+1:

**Pull request merged**